src/vos/vos_dtx.c

```diff
@@ -234,6 +234,20 @@ vos_dtx_table_create(struct vos_pool *pool, struct vos_dtx_table_df *dtab_df)
 	dtab_df->tt_entry_head = UMOFF_NULL;
 	dtab_df->tt_entry_tail = UMOFF_NULL;
 
+	/* For recording the lowest active epoch, we assume that epochs and
+	 * corresponding transactions are generated serially -- i.e. the last
+	 * allocated transaction also has the highest epoch number. This means
+	 * the lowest and highest epochs are guaranteed to be at the head and
+	 * tail of the active list (we only track the tail). If this assumption
+	 * is false, we may need a different approach to keep track of the
+	 * lowest active epoch (either insert new transactions in the correct
+	 * order, or use a min-heap instead of a list).
```

Comment:
Is that true? The transaction epochs are created on different leaders (it may from client when distributed transaction introduced), and server HLCs are not synced, I don't think this order is guaranteed. Fanyong, please correct me if I'm wrong.

What about using 'current_hlc + timeout' as 'stable epoch' for this moment? (as we discussed on the metadata meeting in other day)

Reply:
We cannot guarantee that. Currently, the epoch# for update is generated on leader replica, but different DTXs may have different leader replicas. The epochs among these DTXs are unordered. In the future, after supporting distributed transaction, the epoch# will be generated by client, then it will be more difficult to main the ordering of the DTXs' epoch#

Reply:
agreed, as we discussed in the meeting last week, we should do something like "sync", please check https://github.com/daos-stack/daos/pull/923
briefly, we should have two steps:
1) collect the highest HLC for the snapshot
2) bcast this HLC to all nodes and request them to "sync" all the outstanding DTXs